[slimtensor] Add SizesAndStrides and dimension utility functions for slimtensor usage

Gasoonjia opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #16385
* __->__ #16384
* #16383
* #16382

This diff adds the `SizesAndStrides` class for efficiently storing tensor dimension sizes and strides, along with utility functions for computing element counts and storage requirements.

**Key components:**

1. **`c10/macros/Macros.h`** - Branch prediction hint macros (`SLIMTENSOR_LIKELY/UNLIKELY`) for optimizing hot paths

2. **`util/ArrayRefUtil.h`** - Re-exports ExecuTorch's `ArrayRef`, `IntArrayRef`, and `makeArrayRef`; adds `toVec()` utility for converting ArrayRef to std::vector

3. **`c10/core/SizesAndStrides.h`** - Packed container for tensor sizes and strides:
   - Inline storage for tensors up to 5 dimensions (no heap allocation)
   - Out-of-line heap storage for larger tensors
   - Full copy/move semantics with proper resource management
   - Equality operators and iterators
   - `std::initializer_list` overloads for convenient initialization

4. **`util/SizeUtil.h`** - Size computation utilities:
   - `compute_numel()` - compute total number of elements
   - `compute_contiguous_strides()` - compute row-major contiguous strides
   - `compute_storage_nbytes()` / `compute_storage_nbytes_contiguous()` - compute storage byte requirements

Differential Revision: [D89749336](https://our.internmc.facebook.com/intern/diff/D89749336/)

Gasoonjia committed:
[slimtensor] Add SizesAndStrides and dimension utility functions for slimtensor usage

This diff adds the `SizesAndStrides` class for efficiently storing tensor dimension sizes and strides, along with utility functions for computing element counts and storage requirements.

**Key components:**

1. **`c10/macros/Macros.h`** - Branch prediction hint macros (`SLIMTENSOR_LIKELY/UNLIKELY`) for optimizing hot paths

2. **`util/ArrayRefUtil.h`** - Re-exports ExecuTorch's `ArrayRef`, `IntArrayRef`, and `makeArrayRef`; adds `toVec()` utility for converting ArrayRef to std::vector

3. **`c10/core/SizesAndStrides.h`** - Packed container for tensor sizes and strides:
   - Inline storage for tensors up to 5 dimensions (no heap allocation)
   - Out-of-line heap storage for larger tensors
   - Full copy/move semantics with proper resource management
   - Equality operators and iterators
   - `std::initializer_list` overloads for convenient initialization

4. **`util/SizeUtil.h`** - Size computation utilities:
   - `compute_numel()` - compute total number of elements
   - `compute_contiguous_strides()` - compute row-major contiguous strides
   - `compute_storage_nbytes()` / `compute_storage_nbytes_contiguous()` - compute storage byte requirements

Differential Revision: [D89749336](https://our.internmc.facebook.com/intern/diff/D89749336/)

[ghstack-poisoned]

Gasoonjia committed:
Update on "[slimtensor] Add SizesAndStrides and dimension utility functions for slimtensor usage"


This diff adds the `SizesAndStrides` class for efficiently storing tensor dimension sizes and strides, along with utility functions for computing element counts and storage requirements.

**Key components:**

1. **`c10/macros/Macros.h`** - Branch prediction hint macros (`SLIMTENSOR_LIKELY/UNLIKELY`) for optimizing hot paths

2. **`util/ArrayRefUtil.h`** - Re-exports ExecuTorch's `ArrayRef`, `IntArrayRef`, and `makeArrayRef`; adds `toVec()` utility for converting ArrayRef to std::vector

3. **`c10/core/SizesAndStrides.h`** - Packed container for tensor sizes and strides:
   - Inline storage for tensors up to 5 dimensions (no heap allocation)
   - Out-of-line heap storage for larger tensors
   - Full copy/move semantics with proper resource management
   - Equality operators and iterators
   - `std::initializer_list` overloads for convenient initialization

4. **`util/SizeUtil.h`** - Size computation utilities:
   - `compute_numel()` - compute total number of elements
   - `compute_contiguous_strides()` - compute row-major contiguous strides
   - `compute_storage_nbytes()` / `compute_storage_nbytes_contiguous()` - compute storage byte requirements

Differential Revision: [D89749336](https://our.internmc.facebook.com/intern/diff/D89749336/)

[ghstack-poisoned]

Gasoonjia committed:
Update on "[slimtensor] Add SizesAndStrides and dimension utility functions for slimtensor usage"


This diff adds the `SizesAndStrides` class for efficiently storing tensor dimension sizes and strides, along with utility functions for computing element counts and storage requirements.

**Key components:**

1. **`c10/macros/Macros.h`** - Branch prediction hint macros (`SLIMTENSOR_LIKELY/UNLIKELY`) for optimizing hot paths

2. **`util/ArrayRefUtil.h`** - Re-exports ExecuTorch's `ArrayRef`, `IntArrayRef`, and `makeArrayRef`; adds `toVec()` utility for converting ArrayRef to std::vector

3. **`c10/core/SizesAndStrides.h`** - Packed container for tensor sizes and strides:
   - Inline storage for tensors up to 5 dimensions (no heap allocation)
   - Out-of-line heap storage for larger tensors
   - Full copy/move semantics with proper resource management
   - Equality operators and iterators
   - `std::initializer_list` overloads for convenient initialization

4. **`util/SizeUtil.h`** - Size computation utilities:
   - `compute_numel()` - compute total number of elements
   - `compute_contiguous_strides()` - compute row-major contiguous strides
   - `compute_storage_nbytes()` / `compute_storage_nbytes_contiguous()` - compute storage byte requirements

Differential Revision: [D89749336](https://our.internmc.facebook.com/intern/diff/D89749336/)

[ghstack-poisoned]

manuelcandales reviewed (commented):
don't you need a test_array_ref_util.cpp as well?

Gasoonjia committed:
Update on "[slimtensor] Add SizesAndStrides and dimension utility functions for slimtensor usage"


This diff adds the `SizesAndStrides` class for efficiently storing tensor dimension sizes and strides, along with utility functions for computing element counts and storage requirements.

**Key components:**

1. **`c10/macros/Macros.h`** - Branch prediction hint macros (`SLIMTENSOR_LIKELY/UNLIKELY`) for optimizing hot paths

2. **`util/ArrayRefUtil.h`** - Re-exports ExecuTorch's `ArrayRef`, `IntArrayRef`, and `makeArrayRef`; adds `toVec()` utility for converting ArrayRef to std::vector

3. **`c10/core/SizesAndStrides.h`** - Packed container for tensor sizes and strides:
   - Inline storage for tensors up to 5 dimensions (no heap allocation)
   - Out-of-line heap storage for larger tensors
   - Full copy/move semantics with proper resource management
   - Equality operators and iterators
   - `std::initializer_list` overloads for convenient initialization

4. **`util/SizeUtil.h`** - Size computation utilities:
   - `compute_numel()` - compute total number of elements
   - `compute_contiguous_strides()` - compute row-major contiguous strides
   - `compute_storage_nbytes()` / `compute_storage_nbytes_contiguous()` - compute storage byte requirements

Differential Revision: [D89749336](https://our.internmc.facebook.com/intern/diff/D89749336/)

[ghstack-poisoned]

Gasoonjia committed:
Update on "[slimtensor] Add SizesAndStrides and dimension utility functions for slimtensor usage"


This diff adds the `SizesAndStrides` class for efficiently storing tensor dimension sizes and strides, along with utility functions for computing element counts and storage requirements.

**Key components:**

1. **`c10/macros/Macros.h`** - Branch prediction hint macros (`SLIMTENSOR_LIKELY/UNLIKELY`) for optimizing hot paths

2. **`util/ArrayRefUtil.h`** - Re-exports ExecuTorch's `ArrayRef`, `IntArrayRef`, and `makeArrayRef`; adds `toVec()` utility for converting ArrayRef to std::vector

3. **`c10/core/SizesAndStrides.h`** - Packed container for tensor sizes and strides:
   - Inline storage for tensors up to 5 dimensions (no heap allocation)
   - Out-of-line heap storage for larger tensors
   - Full copy/move semantics with proper resource management
   - Equality operators and iterators
   - `std::initializer_list` overloads for convenient initialization

4. **`util/SizeUtil.h`** - Size computation utilities:
   - `compute_numel()` - compute total number of elements
   - `compute_contiguous_strides()` - compute row-major contiguous strides
   - `compute_storage_nbytes()` / `compute_storage_nbytes_contiguous()` - compute storage byte requirements

Differential Revision: [D89749336](https://our.internmc.facebook.com/intern/diff/D89749336/)

[ghstack-poisoned]

Gasoonjia committed:
Update on "[slimtensor] Add SizesAndStrides and dimension utility functions for slimtensor usage"


This diff adds the `SizesAndStrides` class for efficiently storing tensor dimension sizes and strides, along with utility functions for computing element counts and storage requirements.

**Key components:**

1. **`c10/macros/Macros.h`** - Branch prediction hint macros (`SLIMTENSOR_LIKELY/UNLIKELY`) for optimizing hot paths

2. **`util/ArrayRefUtil.h`** - Re-exports ExecuTorch's `ArrayRef`, `IntArrayRef`, and `makeArrayRef`; adds `toVec()` utility for converting ArrayRef to std::vector

3. **`c10/core/SizesAndStrides.h`** - Packed container for tensor sizes and strides:
   - Inline storage for tensors up to 5 dimensions (no heap allocation)
   - Out-of-line heap storage for larger tensors
   - Full copy/move semantics with proper resource management
   - Equality operators and iterators
   - `std::initializer_list` overloads for convenient initialization

4. **`util/SizeUtil.h`** - Size computation utilities:
   - `compute_numel()` - compute total number of elements
   - `compute_contiguous_strides()` - compute row-major contiguous strides
   - `compute_storage_nbytes()` / `compute_storage_nbytes_contiguous()` - compute storage byte requirements

Differential Revision: [D89749336](https://our.internmc.facebook.com/intern/diff/D89749336/)

[ghstack-poisoned]
